Allow anonymous profile view

ironiridis opened the pull request:
This PR allows unauthenticated viewers to render a profile if it has not opted out of search engine indexing. Would resolve https://github.com/MSPSocial/projects/issues/4#issue-1581332535, which was unfortunately introduced by https://github.com/mastodon/mastodon/pull/19803 .

Considerations:
* This explicitly does not give access to any statuses (ie posts/notes/toots)
* This behavior depends on the search engine indexing opt-out for lack of a better existing pref.
* I considered using the followers-require-approval (the so-called "locked") preference, but the search engine preference felt closer to the spirit of the function, as the whole point of disabling unauthenticated API access is to avoid scrapers.

It would probably be better to put this into a new preference, but personally the whole spectrum of privacy preferences in the software should be represented by something more nuanced than a binary. It seems like a fine-grained privacy permissions system would be better here (with each category of data having selectors for audiences, eg: `only me`, `my followers`, `anybody on my instance`, `anybody on the fediverse`, `anybody on the internet`) but that's a huge project.

Hence this stopgap. (Cc: @lawremipsum, @t54r4n1)

ironiridis committed:
Exempt Emoji API endpoint from auth

When viewing a profile anonymously, the request for custom emoji will show a 401 Unauthorized error to the user. Exempt this endpoint from the authentication requirement.

ironiridis committed:
Move unless criteria into boolean method

It's helpful to be able to ask the parent class whether we would disallow API access based on the current request (and hence whether we know the requesting user) rather than simply based on configuration. This doesn't appear to regress any existing functionality.

ironiridis committed:
Add user-oriented helper methods

This centralizes the core check for the user's preferences into a single helper method, and gives us one place to change the user preference flag if it turns out user_prefers_noindex isn't appropriate. Also adds a convenience error response specific to the user visibility situation.

ironiridis committed:
Allow restricted API lookup if user does not opt-out

ironiridis committed:
Allow restricted API account details if user does not opt-out

ironiridis committed:
Fix front-end to avoid tight request loop on 401 response

ironiridis committed:
Return empty statuses and tags for users visible through restricted API

IanWhitney commented on app/controllers/api/v1/custom_emojis_controller.rb:
```diff
@@ -2,6 +2,7 @@
 
 class Api::V1::CustomEmojisController < Api::BaseController
   skip_before_action :set_cache_headers
+  skip_before_action :require_authenticated_user!
```
Same as previous not on using the `only` param.

IanWhitney commented on app/controllers/api/v1/accounts/featured_tags_controller.rb:
```diff
@@ -3,10 +3,14 @@
   respond_to :json
 
   def index
+    if user_would_block_unauthenticated_api_access?(@account)
+      user_blocks_unauthenticated_api_access and return
+    end
```
```suggestion
    if user_would_block_unauthenticated_api_access?(@account)
      user_blocks_unauthenticated_api_access and return
    end
```

Since you end up having this statement in a bunch of your controllers, you could move it up to the `Api::BaseController` and implement it once.

ironiridis replied:
My very basic Ruby grasp is what led me to doing this this way. I needed `user_blocks_unauthenticated_api_access` to return the JSON object and 401 status, and I wasn't sure how to invoke the render behavior and return from the function. Earlier iterations of this change set had a method that called this, but the calling code would continue and would try to render another result, which resulted in errors in the logs as a response was already being sent. The `... and return` pattern was the only way I could think of to solve this, but I'm very open to alternatives (and to learning!)

IanWhitney replied:
I think the `and return` approach would work still work if you moved this up to a method in the base controller. But maybe not. Controller behavior in Rails is frequently surprising.

IanWhitney commented on app/controllers/api/v1/accounts/featured_tags_controller.rb:
```diff
@@ -17,6 +21,6 @@ def set_account
 
   def set_featured_tags
-    @featured_tags = @account.suspended? ? [] : @account.featured_tags
+    @featured_tags = @account.suspended? || disallow_unauthenticated_api_access? ? [] : @account.featured_tags
```
```suggestion
@featured_tags = if @account.suspended? || disallow_unauthenticated_api_access?
[]
else
@account.featured_tags
end
```

Purely preference here, but I find ternary statements to be hard to read at the best of times. And a ternary with an or it it just seems ripe for misunderstanding.

ironiridis replied:
This is a great suggestion; I forgot that Ruby's `if`'s are expressions and not flow control.

IanWhitney commented on app/controllers/api/v1/accounts/lookup_controller.rb:
```diff
@@ -1,10 +1,14 @@
 # frozen_string_literal: true
 
 class Api::V1::Accounts::LookupController < Api::BaseController
+  skip_before_action :require_authenticated_user!
```
You use the `only` param in the other controllers and I think it's a good idea to do the same here. Yeah, you might skip this action for all controller methods, but at least you're being explicit about it. And you're less likely to get bitten by someone adding a new method that should check for authenticated users.

IanWhitney commented on app/controllers/api/v1/accounts/statuses_controller.rb:
```diff
@@ -18,7 +22,7 @@ def set_account
 
   def load_statuses
-    @account.suspended? ? [] : cached_account_statuses
+    @account.suspended? || disallow_unauthenticated_api_access? ? [] : cached_account_statuses
```
Same suggestion here as on the other ternary

IanWhitney commented on app/controllers/api/base_controller.rb:
```diff
@@ -133,7 +133,16 @@ def set_cache_headers
 
   def disallow_unauthenticated_api_access?
-    ENV['DISALLOW_UNAUTHENTICATED_API_ACCESS'] == 'true' || Rails.configuration.x.whitelist_mode
+    (ENV['DISALLOW_UNAUTHENTICATED_API_ACCESS'] == 'true' || Rails.configuration.x.whitelist_mode) unless current_user
```
```suggestion
    return false if current_user
    ENV['DISALLOW_UNAUTHENTICATED_API_ACCESS'].casecmp?('true' ) || Rails.configuration.x.whitelist_mode
```

Functionally the same, but I think it offers some readability improvements.

By explicitly returning in the first line you're saving some work. 

And `ENV['DISALLOW_UNAUTHENTICATED_API_ACCESS'].casecmp?('true' )` will return a boolean true if the ENV var is "True", "TRUE", "true" or any other combination. Does require Ruby 2.4 or later.

Should you be on < 2.4 I suggest 

```ruby
ENV['DISALLOW_UNAUTHENTICATED_API_ACCESS'].casecmp('true') == 0
```

ironiridis replied:
I agree with the reformat for clearer intention, though I don't want to change the semantics of the env var check here (which are from upstream). Other places that check env vars do so with case sensitive comparisons.

ironiridis committed:
Review fix: Use only directive for all auth skips

IanWhitney's review pointed out that it is more consistent and a safer default to include the "only:" directive for all cases where we skip the user authentication requirement. In particular: if a future public method is added to this controller, it should not be implicitly available pre-auth.

ironiridis committed:
Review fix: Avoid ternary expressions

It seems I let my inner C programmer get the better of me. Per IanWhitney's review, Ruby's if structures are expressions and return values, which make them more suitable than a ternary expression in terms of readability.

ironiridis committed:
Review fix: Restructure for early return

Per IanWhitney's review, the complex logic with the unless clause is pretty clumsy. We can check this condition first and return false immediately, and leave the remaining logic as a simpler statement on its own.

IanWhitney commented on app/controllers/api/v1/accounts/statuses_controller.rb:
```diff
@@ -18,7 +22,10 @@ def set_account
+    if @account.suspended? || disallow_unauthenticated_api_access?
+      []
+    else
+      cached_account_statuses
```
I think this needs one more `end`. The one on line 29 ends the method. But I don't see an `end` for the `if`

ironiridis replied:
Ope, I bet you're right; that's what I get for patching a running EC2 instance and eyeball-copying back to an editor. Will address later today.
